\documentclass[aspectratio=169,12pt,t]{beamer}
\usepackage{graphicx}
\setbeameroption{hide notes}
\setbeamertemplate{note page}[plain]
\usepackage{listings}

\input{header.tex}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% end of header
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% title info
\title{Collaborating reproducibly}
\author{\href{https://kbroman.org}{Karl Broman}}
\institute{Biostatistics \& Medical Informatics \\ Univ.\ Wisconsin{\textendash}Madison}
\date{\href{https://kbroman.org}{\tt \scriptsize \color{foreground} kbroman.org}
\\[-4pt]
\href{https://github.com/kbroman}{\tt \scriptsize \color{foreground} github.com/kbroman}
\\[-4pt]
\href{https://twitter.com/kwbroman}{\tt \scriptsize \color{foreground} @kwbroman}
\\[2pt]
\scriptsize {\lolit Slides:} \href{https://bit.ly/rrcollab}{\tt \scriptsize
  \color{foreground} bit.ly/rrcollab}
}


\begin{document}

% title slide
{
\setbeamertemplate{footline}{} % no page number here
\frame{
  \titlepage

  \vfill \hfill \includegraphics[height=6mm]{Figs/cc-zero.png} \vspace*{-3mm}

  \note{These are slides for a talk I'm going to give at the AAAS
    meeting in Washington, DC, on 17 Feb 2019.

    Source: {\tt https://github.com/kbroman/Talk\_AAAS2019} \\
    Slides: {\tt https://bit.ly/rrcollab}
}
} }


\begin{frame}[fragile,c]{}

\begin{center}
\begin{minipage}[c]{9.3cm}
\begin{semiverbatim}
\lstset{basicstyle=\normalsize}
\begin{lstlisting}[linewidth=9.3cm]
 Karl -- this is very interesting,
 however you used an old version of
 the data (n=143 rather than n=226).

 I'm really sorry you did all that
 work on the incomplete dataset.

 Bruce
\end{lstlisting}
\end{semiverbatim}
\end{minipage}
\end{center}

\note{This is an edited version of an email I got from a collaborator,
  in response to an analysis report that I had sent him.

  I try to always include some brief data summaries at the start of
  such reports. By doing so, he immediately saw that I had an old
  version of the data.

  Because I'd set things up carefully, I could just substitute in the
  newer dataset, type ``{\tt make}'', and get the revised report.

  This is a reproducibility success story. But it took me a long
  time to get to this point.
}
\end{frame}


\begin{frame}[c]{}
\centering
{\Large The results in Table 1 don't seem to \\[12pt]
correspond to those in Figure 2.}

\note{My computational life is not entirely rosy. This is the sort of
  email that will freak me out.}
\end{frame}


\begin{frame}[c]{}
\centerline{\Large In what order do I run these scripts?}

\note{Sometimes the process of data file manipulation and data
  cleaning gets spread across a bunch of scripts that need to be
  executed in a particular order. Will I record this information? Is
  it obvious what script does what?}
\end{frame}



\begin{frame}[c]{}
\centerline{\Large Where did we get this data file?}

\note{Record the provenance of all data or metadata files.}
\end{frame}



\begin{frame}[c]{}
\centerline{\Large Why did I omit those samples?}

\note{I may decide to omit a few samples. Will I record {\nhilit why}
  I omitted those particular samples?}
\end{frame}



\begin{frame}[c]{}
\centerline{\Large How did I make that figure?}

\note{Sometimes, in the midst of a bout of exploratory data analysis,
  I'll create some exciting graph and have a heck of a time
  reproducing it afterwards.}
\end{frame}


\begin{frame}[c]{}
\centerline{\Large Which image goes with which experiment?}

\note{For experimental biologists, it can be tricky to keep track of
  the vast set of images and experiments they perform.}
\end{frame}



\begin{frame}[c]{}
\centerline{\Large ``Your script is now giving an error."}

\note{It was working last week. Well, last month, at least.

How easy is it to go back through that script's history to see where
and why it stopped working?}
\end{frame}



\begin{frame}[c]{}
\centerline{\Large ``The attached is similar to the code we used."}

\note{From an email in response to my request for code used for a
  paper.}
\end{frame}




\begin{frame}[c]{}


\centering
\Large

Reproducible

\bigskip

{\color{lolit} vs.}

\bigskip

\only<1>{Replicable}
\only<2 | handout 0>{Correct}

\note{Computational work is
  {\color{nhilit} reproducible} if one can take the data and code and produce
  the same set of results. {\color{nhilit} Replicable} is more stringent: can
  someone repeat the experiment and get the same results?

  Reproducibility is a minimal standard. That something is
  reproducible doesn't imply that it is correct. The code may have bugs. The
  methods may be poorly behaved. There could be experimental
  artifacts.

  (But reproducibility is probably associated with correctness.)

  Note that some scientists say replicable for what I call
  reproducible, and vice versa.
}
\end{frame}



\begin{frame}[c]{\href{https://kbroman.org/steps2rr}{\tt \color{title}
      kbroman.org/steps2rr}}

  \begin{enumerate}
  \itemsep10pt
  \item Organize your data \& code
  \item Everything with a script
  \item Automate the process {\lolit (GNU Make)}
  \item Turn scripts into reproducible reports
  \item Turn repeated code into functions
  \item Create a package/module
  \item Use version control {\lolit (git/GitHub)}
  \item License your software
  \end{enumerate}


\note{The above are my thoughts on the basic steps to take towards
  full reproducibility.

  Don't try to change every aspect of your workflow all at once.
}
\end{frame}


\begin{frame}[c]{Organize your project}

\begin{center}
\large
\only<1|handout:0>{
File organization and naming \\
are powerful weapons against chaos.
}
\only<2>{
Your closest collaborator is you six months ago, \\
but you don't reply to emails.
}
\only<3>{Have sympathy for your future self.}
\end{center}

\hfill
{\lolit
\only<1|handout:0>{{\textendash} \href{https://jennybryan.org}{Jenny Bryan}}
\only<2>{(paraphrasing \href{https://twitter.com/kcranstn/status/370914072511791104}{Mark Holder})}
\only<3|handout:0>{}
}

\note{The first thing to do is to make your project
  understandable to others (or yourself, later, when you try to figure
  out what it was that you did.

  Segregate all the materials for a project in one directory/folder on
  your harddrive.

  I prefer to separate raw data from processed data, and I put code in
  a separate directory.

  Write {\tt ReadMe} files to explain what's what.
}
\end{frame}



\begin{frame}[fragile,c]{Organize your project}

\begin{center}
\begin{minipage}[c]{10.3cm}
\begin{semiverbatim}
\lstset{basicstyle=\normalsize}
\begin{lstlisting}[linewidth=10.3cm]
RawData/              Notes/
DerivedData/          Refs/

Python/               ReadMe.txt
R/                    ToDo.txt
Ruby/                 Makefile
\end{lstlisting}
\end{semiverbatim}
\end{minipage}
\end{center}

\end{frame}


\begin{frame}<handout:0>[fragile,c]{Organize your project}

\addtocounter{framenumber}{-1}

\begin{center}
\begin{minipage}[c]{10.3cm}
\begin{semiverbatim}
\lstset{basicstyle=\normalsize}
\begin{lstlisting}[linewidth=10.3cm]
0_vcf2db.R
1_prep_geno.R
2_prep_pheno_clin.R
2_prep_pheno_otu.R
3_prep_covar.R
4_prep_analysis_pheno_clin.R
4_prep_analysis_pheno_otu.R
5_scans.R
6_grab_peaks.R
7_find_nearby_peaks.R
\end{lstlisting}
\end{semiverbatim}
\end{minipage}
\end{center}

\end{frame}



\begin{frame}<handout:0>[fragile,c]{Chaos}

\addtocounter{framenumber}{-1}

\begin{center}
\begin{minipage}[c]{11.33cm}
\begin{semiverbatim}
\lstset{basicstyle=\scriptsize}
\begin{lstlisting}[linewidth=11.33cm]
AimeeNullSims/    Deuterium/            Ping/
AimeeResults/     ExtractData4Gary/     Ping2/
AnnotationFiles/  FromAimee/            Ping3/
Brian/            GoldStandard/         Ping4/
Chr6_extrageno/   HumanGWAS/            Play/
Chr6_segdis/      Insulin/              Prdm9/
ChrisPlaisier/    Int2_for_Mark/        RBM_PlasmaUrine_2012-03-08/
Code4Aimee/       Islet_2011-05/        Slco1a6/
CompAnnot/        MappingProbes/        StudyLineupMethods/
CondScans/        MultiProbes/          kidney_chr6.R
D2O_2012-02-14/   NewMap/               pck2_sucla2.R
D2O_cellcycle/    Notes/                penalties.txt
D2Ocorr/          NullSims/             transeQTL4Lude/
Data4Aimee/       NullSims_2009-09-10/
Data4Tram/        PepIns_2012-02-09/
\end{lstlisting}
\end{semiverbatim}
\end{minipage}
\end{center}

\end{frame}




\begin{frame}<handout:0>[c]{No ``{\hilit final}'' in file names}

\vspace*{3mm}

\centering

% comic from http://www.phdcomics.com/comics/archive.php?comicid=1531
\only<1>{\figh{Figs/phd101212s.png}{0.8}}

\end{frame}





\begin{frame}[c]{Reproducible reports}


\vspace*{8mm}

\vspace*{-0.05\textheight}
\figw{Figs/example_Rmd.png}{0.92}
\onslide<2>{
  \vspace*{-0.70\textheight}
  \hspace*{0.06\textwidth}
  \figw{Figs/example_Rmd_source.png}{0.92}
}

\note{I {\nhilit love} R Markdown for making reproducible reports that
  document the full details of my analysis. R Markdown mixes Markdown
  (for light-weight markup of text) and R code chunks; when processed
  with knitr, the R code is executed and results inserted into the
  final document.

  With these informal reports, I seek to fully capture the entirety of
  my data explorations and decisions.

  Python people should look at iPython notebooks.
}
\end{frame}





\begin{frame}[c]{\href{https://kbroman.org/steps2rr}{\tt \color{title} kbroman.org/steps2rr}}

  \begin{enumerate}
  \itemsep10pt
  \item Organize your data \& code
  \item Everything with a script
  \item Automate the process {\lolit (GNU Make)}
  \item Turn scripts into reproducible reports
  \item Turn repeated code into functions
  \item Create a package/module
  \item Use version control {\lolit (git/GitHub)}
  \item License your software
  \end{enumerate}


\note{
    Back to my summary again.
}
\end{frame}






\begin{frame}[c]{Collaboration}

\onslide<2>{
  \bbi
\item Do more, by working in parallel
\item Do more, through diversity of ideas and skills
\item Reproducible pipelines have immediate advantages
\item Tests of reproducibility
\item Code review
  \ei
}

  \note{
    Collaboration has a lot of advantages, including for
    reproducibility efforts.
  }

\end{frame}



\begin{frame}[c]{Challenges in collaborations}

  \bbi
{\only<2->{\lolit }
\item Shared vision?
\item Compromise
\item Coordination
\item Communication
\item Sharing code and data
\item Synchronization
}
\onslide<2->{\item Weakest link?}
  \ei

  \note{
    Collaboration also has challenges.
  }

\end{frame}


\begin{frame}[c]{Genetics of metabolic disease in mice}

  \begin{columns}
    \begin{column}{0.50\textwidth}

\vspace*{12pt}

      \footnotesize
\hspace*{-3mm}
    Alan Attie, UW-Madison, Biochemistry \\[12pt]
\hspace*{-3mm}
    Karl Broman, UW-Madison, Biostat \& Med Info \\[12pt]
\hspace*{-3mm}
    Gary Churchill, Jackson Lab\\[12pt]
\hspace*{-3mm}
    Josh Coon, UW-Madison, Chemistry \\[12pt]
\hspace*{-3mm}
    Federico Rey, UW-Madison, Microbiology \\[12pt]
\hspace*{-3mm}
    Brian Yandell, UW-Madison, Statistics

    \end{column}

    \begin{column}{0.45\textwidth}

    \figw{Figs/cc_founders.png}{0.99}

    \end{column}
  \end{columns}


\end{frame}




\begin{frame}[c]{Diversity outbred mice}

\vspace*{3mm}
\figw{Figs/hs.pdf}{1.0}

\end{frame}



\begin{frame}[c]{Data}


  \begin{columns}
    \begin{column}{0.45\textwidth}
      \bbi
    \item 500 DO mice
      \bi
    \item generations 17--21
    \item high fat, high sugar diet
      \ei
    \item GigaMUGA SNP arrays
      \bi
    \item 140k SNPs
      \ei
    \item Clinical traits
      \bi
    \item Weekly body weight
    \item Glucose tolerance test
    \item Longitudinal serum samples
    \item ex vivo islet insulin secretion
      \ei
      \ei
    \end{column}

    \begin{column}{0.45\textwidth}
      \bbi
    \item Islet gene expression by RNA-seq
    \item Proteins by mass spec
    \item Lipids by mass spec
    \item Gut microbiome
      \bi
    \item 16S RNA
    \item metagenomic data
      \ei
      \ei

    \end{column}
  \end{columns}

\end{frame}


\begin{frame}[c]{Genome scans}


\only<1>{\figw{Figs/attiedo_results_nolegend.pdf}{1.0}}

\only<2>{\figw{Figs/attiedo_results.pdf}{1.0}}

\end{frame}






\begin{frame}[c]{}


\begin{center}
\Large
  {\color{title} Challenges} \\[24pt]
  {\lolit \large (totally hypothetical)}
\end{center}

\end{frame}





\begin{frame}[c]{}

\begin{center}
  \Large

  ``Could we meet to talk about the data file structure?'' \\[36pt]
  \onslide<2->{``No.''}
\end{center}

\end{frame}



\begin{frame}[c]{}

\begin{center}
  \Large


``What the heck is `{\hilit \tt FAD{\textunderscore}NAD SI 8.3{\textunderscore}3.3G}'?''

\end{center}

\end{frame}





\begin{frame}[c]{}

\begin{center}
  \Large


  ``Wait, these results seem to be based \\
  on the older SNP map.''

\end{center}

\end{frame}





\begin{frame}[c]{}

\begin{center}
  \Large


  ``Could you write the methods section?'' \\[36pt]
  ``But I didn't do the work, \\
  and we don't have the code that was used.''



\end{center}

\end{frame}





\begin{frame}[c]{}

\begin{center}
  \Large


``My data analyst has taken a job at Google.''

\end{center}

\end{frame}



\begin{frame}[c]{}

\begin{center}
  \Large


``Could you do these analyses? X said they would, but they're not
  responding to my emails.''

\end{center}

\end{frame}




\begin{frame}[c]{Shared vision}

  \bbi
\item Publication
\item Code \& data sharing
\item Who will do what
\item Timeline
\item Ongoing sharing of methods, results
  \ei

\end{frame}



\begin{frame}[c]{Shared workspace}

  \bbi
\item Project structure
\item Data and metadata formats
\item Software environment
\item Automated sync {\lolit (or it won't happen)}
  \ei

\end{frame}




\begin{frame}[c]{Technology for sharing}

  \bbi
\item Data
  \bi
\item figshare
\item dropbox / box / google drive
  \ei
\item Code
  \bi
\item github / bitbucket
  \ei
\item Pipeline / workflows
  \bi
\item make / drake / snakemake / rake
\item docker containers
\item \href{https://mybinder.org}{\tt mybinder.org} /
  \href{https://wholetale.org}{\tt wholetale.org}
  \ei
  \ei

\end{frame}




\begin{frame}[c]{}

\begin{center}
\large
The most important tool is the {\hilit mindset},\\
when starting, that the end product \\
will be reproducible.
\end{center}

\hfill
{\lolit
{\textendash} \href{https://odin.mdacc.tmc.edu/~kabaggerly/}{Keith Baggerly}
}

\note{So true. Desire for reproducibility is step one.
}
\end{frame}




\begin{frame}[c]{}

\begin{center}
\large
The second-most important tool is {\hilit training}.
\end{center}

\vspace*{12mm}

\hfill
{\lolit
{\textendash} \href{https://kbroman.org}{me} \hspace*{20mm}
}

\note{
}
\end{frame}




\begin{frame}[c]{}

\Large

Slides: \href{https://bit.ly/rrcollab}{\tt bit.ly/rrcollab} \quad
\includegraphics[height=5mm]{Figs/cc-zero.png}

\vspace{10mm}

\href{https://kbroman.org}{\tt kbroman.org}

\vspace{10mm}

\href{https://github.com/kbroman}{\tt github.com/kbroman}

\vspace{10mm}

\href{https://twitter.com/kwbroman}{\tt @kwbroman}


\note{
  Here's where you can find me, as well as the slides for this talk.
}
\end{frame}




\end{document}
